modules/renter/contractor/renew.go

```diff
@@ -0,0 +1,127 @@
+package contractor
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+// managedRenew negotiates a new contract for data already stored with a host.
+// It returns the ID of the new contract. This is a blocking call that
+// performs network I/O.
+func (c *Contractor) managedRenew(fcid types.FileContractID, newEndHeight types.BlockHeight) (types.FileContractID, error) {
+	c.mu.RLock()
+	height := c.blockHeight
+	contract, ok := c.contracts[fcid]
+	host, eok := c.hdb.Host(contract.IP)
+	c.mu.RUnlock()
+	if !ok {
+		return types.FileContractID{}, errors.New("no record of that contract")
+	} else if !eok {
+		return types.FileContractID{}, errors.New("no record of that host")
+	} else if newEndHeight < height {
+		return types.FileContractID{}, errors.New("cannot renew below current height")
+	} else if host.Price.Cmp(maxPrice) > 0 {
+		return types.FileContractID{}, errTooExpensive
+	}
+
+	// get an address to use for negotiation
+	c.mu.Lock()
+	if c.cachedAddress == (types.UnlockHash{}) {
+		uc, err := c.wallet.NextAddress()
+		if err != nil {
+			c.mu.Unlock()
+			return types.FileContractID{}, err
```

Comment:
would much rather see this in an anon func, given that you return while holding a lock. The ability to use `defer Unlock` removes some of the stress associated with locking